### ISSUE TITLE
Integrate Ben's chunking in the new mksurfdat toolchain

### ISSUE THREAD

**Issue by slevisconsulting** [ESCOMP/ctsm#643]
Modify mkmapdata to include calls to OCGIS.

**Comment by bekozi** [comment]
I wanted to get a task list down here for the chunked regridding (see this [board](https://github.com/NCPP/ocgis/projects/1) for more detailed issues). I ordered the tasks by priority.

 * [x] ~~Merge branch for running demo code to the ocgis master.~~ _Done._
 * [ ] Support ocgis installation on Cheyenne.
 * [ ] Support ESMF Unstruct output in ocgis. Currently, the format may only be read.
 * [ ] Track global indices following a spatial subset. Needed to construct a weight file with correct indexing when using a spatial subset intermediary as the source grid.
 * [ ] Verbose output for chunked regridding CLI.

Ping @rsdunlapiv

**Comment by slevisconsulting** [comment]
My current understanding:

1) The ocgis call should replace this command in mkmapdata.sh:

cmd="$mpirun $ESMF_REGRID --ignore_unmapped -s ${INGRID[nfile]} "

By "ocgis call" we mean parts of Ben's bash script that does the spatial subsetting and weight generation:
https://github.com/NCPP/ocgis/blob/i488-full-regrid/examples/do-chunked-rwg-with-ss.sh

I would like an expert's opinion on which parts of Ben's script MUST end up in mkmapdata.sh.

2) I would like an expert's tour of mkmapdata.sh, to help ensure correct translation into python. 

At the Monday meeting we can discuss when to have the tour of mkmapdata.sh to address (1) and (2).

Are (1) and (2) in the correct order here?

**Comment by billsacks** [comment]
From today's discussion: Next step is to just play around with using ocgis, to make sure it will work for our use cases with reasonable memory, time, etc. We can worry about getting a robustly engineered solution as a following step. Can start by just seeing what the cmds are that are done by mkmapdata.

**Comment by slevisconsulting** [comment]
Keeping @billsacks and @ekluzek in the loop:

@negin513 @bekozi and I met today. Negin and Ben worked in Negin’s environment (not using conda) to get Ben's script running. They started troubleshooting from the same error that I got in my attempts: “...requires PIO…”

Now Negin has a script that works with Ben’s source and destination datasets.

Next Negin and Ben tried running the script with a couple of CTSM’s default source and destination datasets and failed. Ben will troubleshoot and let us know in the next day or two.

**Comment by bekozi** [comment]
@negin513, @slevisconsulting: I pushed a fix for the SCRIP grids (ncpp/ocgis#497) to master. Thankfully it was a quick one. I also added some new examples:
* [Chunked regridding with the example SCRIP grids](https://github.com/NCPP/ocgis/blob/d90d5bfc3ba323231b40c05a18cb7188bae74029/examples/CTSM-MapData-Workflow/do-chunked-rwg-ctsm-scrip.sh)
* [Getting a grid's spatial resolution](https://github.com/NCPP/ocgis/blob/d90d5bfc3ba323231b40c05a18cb7188bae74029/examples/CTSM-MapData-Workflow/get_grid_spatial_resolution.py)
* [Converting the grids to a GIS format](https://github.com/NCPP/ocgis/blob/d90d5bfc3ba323231b40c05a18cb7188bae74029/examples/CTSM-MapData-Workflow/convert_to_shapefile.py). [QGIS](https://www.qgis.org/en/site/) is a good option for viewing the shapefiles. ArcGIS is also an option. The plot for one grid looks like [this](https://www.dropbox.com/s/aue1zvqquwb3ctq/Screenshot%20from%202019-05-17%2010-04-25.png?dl=0).

Let me know how it goes!

**Comment by slevisconsulting** [comment]
Working in cheyenne:/gpfs/fs1/work/slevis/ocgis_work

Good news. I ran two scripts successfully:
1) time ./do-this-20190516.sh  ...that Ben and Negin got to work during our meeting last Thursday, which points to Ben's sample files:
- source: ll1280x1280_grid.esmf.nc
- destination: ll1280x1280_grid.esmf.subset.nc
- timing result: 12.956u 1.520s 0:14.87 97.3%
2) time ./do-this-20190520.sh  ...modified to point to two CTSM files:
- source: SCRIPgrid_0.5x0.5_AVHRR_c110228.nc
- destination: SCRIPgrid_4x5_nomask_c110308.nc
- no spatial subsetting involved because both files are global
- timing result: 43.440u 0.964s 0:12.23 363.0%

I will post follow-up tests soon.

**Comment by slevisconsulting** [comment]
Comparing the output from (2) in the previous post to the output from running mkmapdata.sh:

**ncdump -h weights_0520.nc**
netcdf weights_0520 {
dimensions:
        n_s = 161480 ;
variables:
        int row(n_s) ;
        int col(n_s) ;
        double S(n_s) ;
}

**ncdump -h map_0.5x0.5_AVHRR_to_4x5_nomask_aave_da_c190516.nc**
netcdf map_0.5x0.5_AVHRR_to_4x5_nomask_aave_da_c190516 {
dimensions:
        n_a = 259200 ;
        n_b = 3312 ;
        n_s = 108026 ;
        nv_a = 4 ;
        nv_b = 4 ;
        num_wgts = 1 ;
        src_grid_rank = 2 ;
        dst_grid_rank = 2 ;
variables:
        int src_grid_dims(src_grid_rank) ;
        int dst_grid_dims(dst_grid_rank) ;
        double yc_a(n_a) ;
                yc_a:units = "degrees" ;
        double yc_b(n_b) ;
                yc_b:units = "degrees" ;
        double xc_a(n_a) ;
                xc_a:units = "degrees" ;
        double xc_b(n_b) ;
                xc_b:units = "degrees" ;
        double yv_a(n_a, nv_a) ;
                yv_a:units = "degrees" ;
        double xv_a(n_a, nv_a) ;
                xv_a:units = "degrees" ;
        double yv_b(n_b, nv_b) ;
                yv_b:units = "degrees" ;
        double xv_b(n_b, nv_b) ;
                xv_b:units = "degrees" ;
        int mask_a(n_a) ;
                mask_a:units = "unitless" ;
        int mask_b(n_b) ;
                mask_b:units = "unitless" ;
        double area_a(n_a) ;
                area_a:units = "square radians" ;
        double area_b(n_b) ;
                area_b:units = "square radians" ;
        double frac_a(n_a) ;
                frac_a:units = "unitless" ;
        double frac_b(n_b) ;
                frac_b:units = "unitless" ;
        int col(n_s) ;
        int row(n_s) ;
        double S(n_s) ;

// global attributes:
                :title = "ESMF Offline Regridding Weight Generator" ;
                :normalization = "destarea" ;
                :map_method = "Conservative remapping" ;
                :ESMF_regrid_method = "First-order Conservative" ;
                :conventions = "NCAR-CSM" ;
                :domain_a = "/glade/p/cesm/cseg/inputdata/lnd/clm2/mappingdata/grids/SCRIPgrid_0.5x0.5_AVHRR_c110228.nc" ;
                :domain_b = "/glade/p/cesm/cseg/inputdata/lnd/clm2/mappingdata/grids/SCRIPgrid_4x5_nomask_c110308.nc" ;
                :grid_file_src = "/glade/p/cesm/cseg/inputdata/lnd/clm2/mappingdata/grids/SCRIPgrid_0.5x0.5_AVHRR_c110228.nc" ;
                :grid_file_dst = "/glade/p/cesm/cseg/inputdata/lnd/clm2/mappingdata/grids/SCRIPgrid_4x5_nomask_c110308.nc" ;
                :CVS_revision = "7.0.0" ;
                :history = "Thu May 16 10:10:03 2019: ncatted -a history,global,a,c,/glade/u/apps/ch/opt/esmf/7.0.0-ncdfio/intel/17.0.1/bin/binO/Linux.intel.64.mpiuni.default/ESMF_RegridWeightGen map_0.5x0.5_AVHRR_to_4x5_nomask_aave_da_c190516.nc\n",
                        "/glade/u/apps/ch/opt/esmf/7.0.0-ncdfio/intel/17.0.1/bin/binO/Linux.intel.64.mpiuni.default/ESMF_RegridWeightGen" ;
                :NCO = "netCDF Operators version 4.7.4 (http://nco.sf.net)" ;
                :hostname = "r5i2n31" ;
                :logname = "slevis" ;

The S variable from (2)
![S_weights_0520](https://user-images.githubusercontent.com/35322900/58137302-a310ea00-7be6-11e9-9182-03b1027ba426.JPG)

The S variable from mkmapdata
![S_map_0 5x0 5_AVHRR_to_4x5_nomask_aave_da](https://user-images.githubusercontent.com/35322900/58137310-ae641580-7be6-11e9-9134-abb83fd95b80.JPG)

The lines in these plots are densely packed dots. There is only one S value per n_s.

**Comment by bekozi** [comment]
@slevisconsulting Is it possible to include the regridding commands that generated these weight files? It would be useful to verify the same flags are being used. Thanks!

**Comment by slevisconsulting** [comment]
It took me and @bekozi a few iterations, but we have a script that works:
`qinteractive -l select=1:ncpus=4:mpiprocs=4`
`time ./do-this-20190624.sh`
This now returns the same weights (to within roundoff) as mkmapdata.sh in test (2) above (plot posted [here](https://github.com/ESCOMP/ctsm/issues/643#issuecomment-494598239) with these timing results:
19.597u 1.458s 0:07.23 291.0%

**Comment by slevisconsulting** [comment]
Good news:
Tests 3 and 4 (previous post) now return the same weights as clm's mkmapdata tool.

**Comment by slevisconsulting** [comment]
Results from new tests.
PASS means that ocgis and mkmapdata return the same weights to within round-off.

5) `time ./do-this-20190808.sh`  requesting 40 chunks  **FAIL**
returns S vector with 9202808 elements compared to mkmapdata's 9171428...

6) `time ./do-this-20190808b.sh`  requesting 40 chunks  **PASS**
216.814u 5.131s 0:57.66 384.9%

7) `time ./do-this-20190808c.sh`  requesting 40 chunks  **PASS**
561.639u 7.694s 2:25.21 392.0%

8) `time ./do-this-20190808d.sh`  requesting 40 chunks  **PENDING**

**Comment by billsacks** [comment]
According to @ekluzek , this issue blocks #806

**Comment by slevisconsulting** [comment]
@ekluzek pls confirm whether I understand the dependencies correctly:
 
@billsacks I think it's the other way around:

#643 depends on #806 
(and #806 depends on https://github.com/NCPP/ocgis/issues/494)

**Comment by ekluzek** [comment]
No, I would describe #806 as depending on this one, because you want to use the subsetting that's available in OCGIS to accomplish #806. So the first thing to do is to move mkmapdata.sh to use OCGIS, and then add the ability to also use it to do the subsetting. There are going to be changes that will be needed to mkmapdata.sh in order for it to use OCGIS either for subsetting or making maps, so you might as well do that first, and then adding the subsetting functionality to it won't be hard to do.

**Comment by ekluzek** [comment]
So a way to look at it is this. This issue is refactoring mkmapdata.sh to use OCGIS. Doing that will make it easier to add subsetting (using OCGIS) to it. So this is a refactoring that enables adding a new feature easier.

**Comment by slevisconsulting** [comment]
**Meeting with @ekluzek** 
...last Friday at which Sam and Erik started merging Sam's script that runs ocgis with mkmapdata.sh.

Currently @slevisconsulting has
`export PYTHONPATH="/glade/work/slevis/git_ocgis/ocgis/src/:/glade/work/slevis/git_esmf/esmf/src/addon/ESMPy/src/"`
in his script that runs ocgis.

Instead, we will need to add ocgis to ctsm as an external.
See Externals_CLM.cfg  under  PTCLM for a template.
When we bring ocgis to master, @bekozi will need to make a new release tag that includes the latest fixes.

Something different will need to happen with ESMPy. Not clear what, yet.

**Comment by ekluzek** [comment]
I've added another issue to OCGIS to add more metadata to their mapping files...

https://github.com/NCPP/ocgis/issues/506

**Comment by billsacks** [comment]
Based on discussion in #645 we actually don't want ocgis as an external: instead, this will be installed on the system via conda.

**Comment by slevisconsulting** [comment]
@billsacks @ekluzek 
No news on issue https://github.com/NCPP/ocgis/issues/506 
which currently blocks this issue.

**Comment by ekluzek** [comment]
With @mvertens work on making mksurfdata run in parallel, I think this issue becomes a WONTFIX. Once, we've completely validated the parallel mksurfdata we should close this.
